libr/bin/format/objc/mach0_classes.c

```diff
@@ -0,0 +1,734 @@
+
+	if (left < sizeof(Mach0_struct_(SClass))) {
+		r_buf_read_at(arch->buf, r, (ut8 *)&c, left);
+	} else {
+		r_buf_read_at(arch->buf, r, (ut8 *)&c, sizeof(Mach0_struct_(SClass)));
+	}
+
+	processed_class->addr = c.isa;
+	Mach0_class_(get_class_ro_t)((c.data) & ~0x3, arch,\
+								&is_meta_class, processed_class);
+}
+
+///////////////////////////////////////////////////////////////////////////////
+static void Mach0_class_(__r_bin_class_free)(RBinClass *p) {
+	RListIter *iter = NULL;
+	RBinSymbol *symbol = NULL;
+	RBinField *field = NULL;
+
+	if (!p) {
+		return;
+	}
+
+	r_list_foreach(p->methods, iter, symbol) {
+		R_FREE(symbol);
+	}
+
+	r_list_foreach(p->fields, iter, field) {
+		R_FREE(field);
+	}
+
+	r_list_free(p->methods);
+	r_list_free(p->fields);
+	r_bin_class_free (p);
+}
+
+RList* Mach0_class_(get_classes)(RBinFile *arch)
+{
+	RList/*<RBinClass>*/ *ret = NULL;
+
+	RBinClass *processed_class = NULL;
+
+	RList *sctns = NULL;
+	RListIter *iter = NULL;
+	RBinSection *s = NULL;
+	ut32 i = 0, size = sizeof(ut32);
+	mach0_ut p;
+	ut32 left;
+	ut8 isFound = 0;
+
+	ut64 numOfUnnamedClass = 0;
+
+	if (!arch || !arch->o || !arch->o->bin_obj)
+		return NULL;
+
+	// searching of section with name __objc_classlist
+	sctns = r_bin_plugin_mach.sections(arch);
+	if (sctns == NULL) {
+		eprintf("there is no sections\n");
+		return NULL;
+	}
+
+	r_list_foreach(sctns, iter, s) {
+		if (strstr(s->name, "__objc_classlist") != NULL) {
+			isFound = 1;
+			break;
+		}
+	}
+
+	if (!isFound) {
+		eprintf("there is no section __objc_classlist\n");
+		goto get_classes_error;
+	}
+	// end of seaching of section with name __objc_classlist
+
+	if (!(ret = r_list_new ())) {
+		eprintf("RList<RBinClass> allocation error\n");
+		goto get_classes_error;
+	}
+
+	ret->free = (RListFree) Mach0_class_(__r_bin_class_free);
+
+	// start of getting information about each class in file
+	for (i = 0; i < s->size; i+= sizeof(mach0_ut)) {
+		if (!(processed_class = R_NEW0 (RBinClass))) {
+			eprintf("RBinClass allocation error\n");
+			goto get_classes_error;
+		}
+
+		if (!(processed_class->methods = r_list_new())) {
+			eprintf("RList<RBinField> allocation error\n");
+			goto get_classes_error;
+		}
+
+		if (!(processed_class->fields = r_list_new())) {
+			eprintf("RList<RBinSymbol> allocation error\n");
+			goto get_classes_error;
+		}
+
+		memset(&p, '\0', sizeof(mach0_ut));
+
+		left = s->size - i;
+		size = left < sizeof(mach0_ut) ? left : sizeof(mach0_ut);
+
+		r_buf_read_at(arch->buf, s->vaddr - r_bin_plugin_mach.baddr(arch) + i, (ut8 *)&p, size);
+
+		Mach0_class_(get_class_t)(p, arch, processed_class);
+
+		if (!processed_class->name) {
+			st8 *tmp = 0;
+			size_t len = 0;
+
+			tmp = r_str_newf("%s%llu", "UnnamedClass", numOfUnnamedClass);
+			len = strlen(tmp);
+			processed_class->name = malloc(len + 1);
+			if (!processed_class->name) {
+				goto get_classes_error;
+			}
+			memset(processed_class->name, 0, len + 1);
```

Comment:
I think you don't need a memset here, because you are doing a strcpy just in the next line.